# Housing Rental Analysis for San Francisco

In [28]:
# Initial imports
import pandas as pd
import hvplot.pandas
from pathlib import Path

In [29]:
# Create a dataframe from sfo_neighborhoods_census_data.csv
sfo_data_df = pd.read_csv(Path("./Resources/sfo_neighborhoods_census_data.csv"))

# Review dataframe to verify
sfo_data_df

year            neighborhood  sale_price_sqr_foot  housing_units  \
0    2010            Alamo Square           291.182945         372560   
1    2010              Anza Vista           267.932583         372560   
2    2010                 Bayview           170.098665         372560   
3    2010        Buena Vista Park           347.394919         372560   
4    2010        Central Richmond           319.027623         372560   
..    ...                     ...                  ...            ...   
392  2016          Telegraph Hill           903.049771         384242   
393  2016              Twin Peaks           970.085470         384242   
394  2016  Van Ness/ Civic Center           552.602567         384242   
395  2016       Visitacion Valley           328.319007         384242   
396  2016           Westwood Park           631.195426         384242   

     gross_rent  
0          1239  
1          1239  
2          1239  
3          1239  
4          1239  
..          ...  
392        4390  
393        4390  
394        4390  
395        4390  
396        4390  

[397 rows x 5 columns]

---

## Calculate and Plot the Housing Units per Year

In [30]:
# Group by year and average the results
housing_units_by_year = sfo_data_df.groupby("year").mean()

# Review dataframe to verify
housing_units_by_year

sale_price_sqr_foot  housing_units  gross_rent
year                                                
2010           369.344353       372560.0      1239.0
2011           341.903429       374507.0      1530.0
2012           399.389968       376454.0      2324.0
2013           483.600304       378401.0      2971.0
2014           556.277273       380348.0      3528.0
2015           632.540352       382295.0      3739.0
2016           697.643709       384242.0      4390.0

In [31]:
# Create a bar chart of the available housing units for each year
housing_units_by_year.hvplot.bar(
    x="year", 
    y="housing_units", 
    ylim=(350000, 400000), 
    title="Housing Units in San Francisco from 2010 to 2016", 
    xlabel="Year", 
    ylabel="Housing Units",
    height=500,
    width=1000
).opts(
    yformatter="%.0f", 
    color="lightblue", 
    hover_color="blue"
)

:Bars   [year]   (housing_units)

**Question: What’s the overall trend in housing units over the period being analyzed?**

The overall trend in the amount of housing units over the time analyzed is a steady increase in the amount available each year.

---

## Calculate and Plot the Average Sale Prices per Square Foot

In [32]:
# Create a numerical aggregation by grouping the data by year and averaging the results
prices_square_foot_by_year = sfo_data_df.groupby("year").mean()

# Review dataframe to verify
prices_square_foot_by_year

sale_price_sqr_foot  housing_units  gross_rent
year                                                
2010           369.344353       372560.0      1239.0
2011           341.903429       374507.0      1530.0
2012           399.389968       376454.0      2324.0
2013           483.600304       378401.0      2971.0
2014           556.277273       380348.0      3528.0
2015           632.540352       382295.0      3739.0
2016           697.643709       384242.0      4390.0

**Question: What is the lowest gross rent reported for the years included in the DataFrame?**

The lowest gross rent reported for the years analyzed is 2010. The gross rent continued to increase each year following.

In [33]:
# Drop housing_units from dataframe
prices_square_foot_by_year = prices_square_foot_by_year.drop(columns="housing_units")

# Review dataframe to verify
prices_square_foot_by_year

sale_price_sqr_foot  gross_rent
year                                 
2010           369.344353      1239.0
2011           341.903429      1530.0
2012           399.389968      2324.0
2013           483.600304      2971.0
2014           556.277273      3528.0
2015           632.540352      3739.0
2016           697.643709      4390.0

In [34]:
# Plot prices_square_foot_by_year
prices_square_foot_by_year.hvplot(
    xlabel="Year",
    ylabel="Gross Rent / Sale Price Per Square Foot",
    title="Sale Price Per Square Foot and Average Gross Rent - San Francisco - 2010 - 2016",
    height=500,
    width=1000
)

:NdOverlay   [Variable]
   :Curve   [year]   (value)

**Question: Did any year experience a drop in the average sale price per square foot compared to the previous year?**

The only year to experience a drop in the average sale price per square foot compared to the previous year was 2011. The price dropped from $369.34 to $341.90.

**Question: If so, did the gross rent increase or decrease during that year?**

Despite the decrease in average sale price per square foot, the gross rent continued to increase during that year.

---

## Compare the Average Sale Prices by Neighborhood

In [35]:
# Group by year and neighborhood and then create a new dataframe of the mean values
prices_by_year_by_neighborhood = sfo_data_df.groupby(["year", "neighborhood"]).mean()

# Review the DataFrame
prices_by_year_by_neighborhood

sale_price_sqr_foot  housing_units  gross_rent
year neighborhood                                                          
2010 Alamo Square                     291.182945       372560.0      1239.0
     Anza Vista                       267.932583       372560.0      1239.0
     Bayview                          170.098665       372560.0      1239.0
     Buena Vista Park                 347.394919       372560.0      1239.0
     Central Richmond                 319.027623       372560.0      1239.0
...                                          ...            ...         ...
2016 Telegraph Hill                   903.049771       384242.0      4390.0
     Twin Peaks                       970.085470       384242.0      4390.0
     Van Ness/ Civic Center           552.602567       384242.0      4390.0
     Visitacion Valley                328.319007       384242.0      4390.0
     Westwood Park                    631.195426       384242.0      4390.0

[397 rows x 3 columns]

In [36]:
# Drop housing_units from dataframe
prices_by_year_by_neighborhood = prices_by_year_by_neighborhood.drop(columns="housing_units")

# Review dataframe to verify
prices_by_year_by_neighborhood

sale_price_sqr_foot  gross_rent
year neighborhood                                           
2010 Alamo Square                     291.182945      1239.0
     Anza Vista                       267.932583      1239.0
     Bayview                          170.098665      1239.0
     Buena Vista Park                 347.394919      1239.0
     Central Richmond                 319.027623      1239.0
...                                          ...         ...
2016 Telegraph Hill                   903.049771      4390.0
     Twin Peaks                       970.085470      4390.0
     Van Ness/ Civic Center           552.602567      4390.0
     Visitacion Valley                328.319007      4390.0
     Westwood Park                    631.195426      4390.0

[397 rows x 2 columns]

In [37]:
# Use hvplot to create an interactive line plot of the average price per square foot
prices_by_year_by_neighborhood.hvplot(
    x="year",
    groupby="neighborhood",
    xlabel="Year",
    ylabel="Gross Rent / Sale Price Per Square Foot",
    title="Sale Price Per Square Foot and Average Gross Rent - By Neighborhood - 2010 - 2016",
    height=500,
    width=1000
)

:DynamicMap   [neighborhood]
   :NdOverlay   [Variable]
      :Curve   [year]   (value)

**Question: For the Anza Vista neighborhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012?**

For the Anza Vista neighborhood, the average sales price per square foot for 2016 was lower than the price for 2012.

---

## Build an Interactive Neighborhood Map

In [38]:
# Load neighborhoods coordinates data
neighborhood_locations_df = pd.read_csv(
    Path("./Resources/neighborhoods_coordinates.csv"),
    index_col = "Neighborhood"
)

# Review the DataFrame
neighborhood_locations_df

Lat         Lon
Neighborhood                             
Alamo Square        37.791012 -122.402100
Anza Vista          37.779598 -122.443451
Bayview             37.734670 -122.401060
Bayview Heights     37.728740 -122.410980
Bernal Heights      37.728630 -122.443050
...                       ...         ...
West Portal         37.740260 -122.463880
Western Addition    37.792980 -122.435790
Westwood Highlands  37.734700 -122.456854
Westwood Park       37.734150 -122.457000
Yerba Buena         37.792980 -122.396360

[73 rows x 2 columns]

In [39]:
# Calculate the mean values for each neighborhood
all_neighborhood_info_df = sfo_data_df.groupby("neighborhood").mean()

# Review the resulting DataFrame
all_neighborhood_info_df

year  sale_price_sqr_foot  housing_units  \
neighborhood                                                          
Alamo Square        2013.000000           366.020712      378401.00   
Anza Vista          2013.333333           373.382198      379050.00   
Bayview             2012.000000           204.588623      376454.00   
Bayview Heights     2015.000000           590.792839      382295.00   
Bernal Heights      2013.500000           576.746488      379374.50   
...                         ...                  ...            ...   
West Portal         2012.250000           498.488485      376940.75   
Western Addition    2012.500000           307.562201      377427.50   
Westwood Highlands  2012.000000           533.703935      376454.00   
Westwood Park       2015.000000           687.087575      382295.00   
Yerba Buena         2012.500000           576.709848      377427.50   

                     gross_rent  
neighborhood                     
Alamo Square        2817.285714  
Anza Vista          3031.833333  
Bayview             2318.400000  
Bayview Heights     3739.000000  
Bernal Heights      3080.333333  
...                         ...  
West Portal         2515.500000  
Western Addition    2555.166667  
Westwood Highlands  2250.500000  
Westwood Park       3959.000000  
Yerba Buena         2555.166667  

[73 rows x 4 columns]

In [40]:
# Using the Pandas `concat` function, join the 
# neighborhood_locations_df and the all_neighborhood_info_df DataFrame
# The axis of the concatenation is "columns".
# The concat function will automatially combine columns with
# identical information, while keeping the additional columns.
all_neighborhoods_df = pd.concat(
    [neighborhood_locations_df, all_neighborhood_info_df], 
    axis="columns",
    sort=False
)

# Review the resulting DataFrame
all_neighborhoods_df

Lat         Lon         year  sale_price_sqr_foot  \
Alamo Square     37.791012 -122.402100  2013.000000           366.020712   
Anza Vista       37.779598 -122.443451  2013.333333           373.382198   
Bayview          37.734670 -122.401060  2012.000000           204.588623   
Bayview Heights  37.728740 -122.410980  2015.000000           590.792839   
Bernal Heights   37.728630 -122.443050          NaN                  NaN   
...                    ...         ...          ...                  ...   
Yerba Buena      37.792980 -122.396360  2012.500000           576.709848   
Bernal Heights         NaN         NaN  2013.500000           576.746488   
Downtown               NaN         NaN  2013.000000           391.434378   
Ingleside              NaN         NaN  2012.500000           367.895144   
Outer Richmond         NaN         NaN  2013.000000           473.900773   

                 housing_units   gross_rent  
Alamo Square          378401.0  2817.285714  
Anza Vista            379050.0  3031.833333  
Bayview               376454.0  2318.400000  
Bayview Heights       382295.0  3739.000000  
Bernal Heights             NaN          NaN  
...                        ...          ...  
Yerba Buena           377427.5  2555.166667  
Bernal Heights        379374.5  3080.333333  
Downtown              378401.0  2817.285714  
Ingleside             377427.5  2509.000000  
Outer Richmond        378401.0  2817.285714  

[77 rows x 6 columns]

In [41]:
# Call the dropna function to remove any neighborhoods that do not have data
all_neighborhoods_df = all_neighborhoods_df.reset_index().dropna()

# Rename the "index" column as "Neighborhood" for use in the Visualization
all_neighborhoods_df = all_neighborhoods_df.rename(columns={"index": "Neighborhood"})

# Review the resulting DataFrame
all_neighborhoods_df

Neighborhood        Lat         Lon         year  \
0         Alamo Square  37.791012 -122.402100  2013.000000   
1           Anza Vista  37.779598 -122.443451  2013.333333   
2              Bayview  37.734670 -122.401060  2012.000000   
3      Bayview Heights  37.728740 -122.410980  2015.000000   
5     Buena Vista Park  37.768160 -122.439330  2012.833333   
..                 ...        ...         ...          ...   
68         West Portal  37.740260 -122.463880  2012.250000   
69    Western Addition  37.792980 -122.435790  2012.500000   
70  Westwood Highlands  37.734700 -122.456854  2012.000000   
71       Westwood Park  37.734150 -122.457000  2015.000000   
72         Yerba Buena  37.792980 -122.396360  2012.500000   

    sale_price_sqr_foot  housing_units   gross_rent  
0            366.020712      378401.00  2817.285714  
1            373.382198      379050.00  3031.833333  
2            204.588623      376454.00  2318.400000  
3            590.792839      382295.00  3739.000000  
5            452.680591      378076.50  2698.833333  
..                  ...            ...          ...  
68           498.488485      376940.75  2515.500000  
69           307.562201      377427.50  2555.166667  
70           533.703935      376454.00  2250.500000  
71           687.087575      382295.00  3959.000000  
72           576.709848      377427.50  2555.166667  

[69 rows x 7 columns]

In [42]:
# Create a plot to analyze neighborhood info
all_neighborhoods_df.hvplot.points(
    'Lon', 
    'Lat',
    geo=True,
    tiles="OSM",
    size="sale_price_sqr_foot",
    color="gross_rent",
    frame_width=700,
    frame_height=500,
    title="Sale Price Per Square Foot and Average Gross Rent By Neighborhood"
)

:Overlay
   .Tiles.I  :Tiles   [x,y]
   .Points.I :Points   [Lon,Lat]   (gross_rent,sale_price_sqr_foot)

**Question: Which neighborhood has the highest gross rent, and which has the highest sale price per square foot?**

The neighborhood with the highest gross rent is Westwood Park and the neighborhood with the highest sales price per square foot is Union Square District.

## Compose Your Data Story

Based on the visualizations that you have created, compose a data story that synthesizes your analysis by answering the following questions:

**Question:**  How does the trend in rental income growth compare to the trend in sales prices? Does this same trend hold true for all the neighborhoods across San Francisco?

**Answer:** # YOUR ANSWER HERE

**Question:** What insights can you share with your company about the potential one-click, buy-and-rent strategy that they're pursuing? Do neighborhoods exist that you would suggest for investment, and why?

**Answer:** # YOUR ANSWER HERE